### DISCUSSION TITLE
How to get CSS container queries to work together with flex justify-between?

### DISCUSSION THREAD

**Discussion by martinszeltins** [tailwindlabs/tailwindcss#10417] [Help]
I have this simple flexbox with 2 children. I want to make my 2nd child a css container and then style its children based on the container size. The problem is that as soon as I add the `@container` class to my 2nd child it disappears. What am I doing wrong?

Here is my code:

````
<div class="flex justify-between text-3xl">
    <div class="bg-blue-200">First</div>

    <div class="@container bg-red-300">
        <div class="@lg:underline">
            Second
        </div>
    </div>
</div>
````

This is how it SHOULD look:
![image](https://user-images.githubusercontent.com/34019878/214815305-0c66b5e3-1bcb-46bf-96bb-096067931798.png)

But this is how it looks with `@container`
![image](https://user-images.githubusercontent.com/34019878/214815445-fc19ae53-b8eb-4859-accb-6be3ef9d0215.png)

**Comment by wongjn** [comment]
The `@container` class applies `container-type: inline-size`, which [as per MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/container-type#values):

> Establishes a query container for dimensional queries on the [inline axis](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Logical_Properties/Basic_concepts#block_and_inline_dimensions) of the container. Applies layout, style, and inline-size containment to the element.

Then if we follow the thread to what inline-size containment is, again [from MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/contain#inline-size):
> Inline size containment is applied to the element. The inline size of the element can be computed in isolation, ignoring the child elements.

The <q>ignoring the child elements</q> clause is important here – the red element's width is entirely dependent on the content, in this case, the <q>Second</q> text. `container-type: inline-size` ignores this and thus the element is `0` width and disappears. You will want to look at establishing a size for the `@container` element that does not rely on its content.

**Reply by martinszeltins** [reply]
> You will want to look at establishing a size for the @container element that does not rely on its content.

How is this even possibe? How do I do that? That defeats the whole purpose of using `flex` and `justify-between`. I don't want to set a hard-coded `width` for my 2nd flex child. That doesn't make sense. I'm confused here...

**Reply by martinszeltins** [reply]
@wongjn 

> element's width is entirely dependent on the content

Isn't that the case in most situations? I mean I almost never set a hard width for any elements (or very rarely). Does it mean that in order to use CSS container queries you need to always provide a hard with for the "container" element??

**Reply by wongjn** [reply]
This limitation is in place otherwise you could establish a recursive loop. Consider the following markup:

```html
<div class="flex">
  <div class="@container">
    <p class="w-96 @xs:w-px">
```

The browser could make the `<p>` element `24rem` in width. However, if the `@container` size *would* establish it's inline-size from its content, then it would be `24rem` in width too. Then, the `@xs:w-px` would match and the `<p>` element would now be `1px` in width. Then the `@container` size would again resize down to `1px`, and then we're back to the beginning again.

**Reply by wongjn** [reply]
> @wongjn
> 
> > element's width is entirely dependent on the content
> 
> Isn't that the case in most situations? I mean I almost never set a hard width for any elements (or very rarely). Does it mean that in order to use CSS container queries you need to always provide a hard with for the "container" element??

You don't need to provide a "hard width" (i.e. via the `width` property) but the `@container` should have some width that is not dependent on it's content. The main incompatibility here is the nature of flex items "shrink-wrapping". For a really simple example, consider this markup:

```html
<div class="@container"></div>
```

Here, the `@container` is a block element, and thus, its width is implicitly `100%`. Notice how we have not defined a "hard" width on it. Also notice how its width is not dependent on its content.

A slightly more complex example:

```html
<div class="@container max-w-lg"></div>
```
Now, we have limited the maximum width of the `@container`, but we have still not defined a "hard" width on it and its width is still not dependent on its content.

And lastly, perhaps something more relevant:

```html
<div class="grid grid-cols-[1fr_3fr] max-w-lg mx-auto p-5">
  <aside class="@container">
    <!-- … -->
  </aside>
  <main class="@container">
    <!-- … -->
  </main>
</div>
```
A layout with a left sidebar. Again, no defined "hard" width and the widths are not dependent on its content.

**Reply by martinszeltins** [reply]
I see, so in order to use css container queries, the width CANNOT be dependent on the content (like my flex example)?

**Reply by wongjn** [reply]
Indeed.

**Reply by martinszeltins** [reply]
With this `<div class="@container max-w-lg"></div>` - you defined the "max width" but how does this not make the width still dependant on its content? What is its width then _(until it reaches max width)_?

**Reply by wongjn** [reply]
The `@container` is a block element, and thus, its width is implicitly `100%` and will be the width of its parent until it is limited to the value defined for `max-width` by the `max-w-lg` rule.

**Reply by martinszeltins** [reply]
For example, this still does not work, even with `max-w-lg`:

````
<div class="flex justify-between text-3xl">
    <div class="bg-blue-200">First</div>

    <div class="@container bg-red-300 max-w-lg">
        <div class="@lg:underline">
            Second
        </div>
    </div>
</div>
````

**Reply by wongjn** [reply]
Yup, as mentioned earlier, it is due to the main incompatibility of the nature of child elements in flex layout "shrink-wrapping":

```html
<div></div>
```
The `<div>` here would be `100%` width.

```html
<section class="flex">
  <div></div>
</section>
```
The `<div>` here would be `0` width.

**Reply by martinszeltins** [reply]
So if I am understanding it correctly then I need to do something like this?

````
<div class="flex justify-between text-3xl">
    <div class="bg-blue-200">First</div>

    <div class="@container bg-red-300 w-1/5">
        <div class="@lg:underline">
            Second
        </div>
    </div>
</div>
````

**Reply by wongjn** [reply]
That would work, yes.

**Reply by martinszeltins** [reply]
True, it would work but then it creates this large empty space....

![image](https://user-images.githubusercontent.com/34019878/214825848-07659a19-3a25-4b27-b3e5-3cadf6463126.png)
